# Hyperparameter Optimization using H2O

This notebook compares the performance of H2O’s grid search, randomized grid search, and AutoML for hyperparameter optimization of the H2ORandomForestEstimator model.

## Overview
The key steps involve performing grid search, randomized grid search, and using H2O AutoML to find the best hyperparameters and model performance.

## Procedure
- **Grid Search**:
  - Perform grid search for H2ORandomForestEstimator with specified hyperparameters.
  - Display sorted grid results and evaluate the best model's performance.
- **Randomized Grid Search**:
  - Perform hyperparameter optimization using randomized grid search.
  - Display sorted results and evaluate the best model's performance.
- **H2O AutoML**:
  - Use H2O AutoML to find the best deep learning model for the classification task.
  - Display the leaderboard and evaluate the best model's performance.

References:
- [H2O Documentation](https://docs.h2o.ai)
- [allyears2k Headers Dataset](https://www.h2o.ai/resources/datasets/allyears2k_headers.zip)

In [5]:
import h2o
from h2o.estimators.gbm import H2OGradientBoostingEstimator
from h2o import estimators
h2o.init()

# import the airlines dataset:
# This dataset is used to classify whether a flight will be delayed 'YES' or not "NO"
# original data can be found at http://www.transtats.bts.gov/
airlines= h2o.import_file("https://s3.amazonaws.com/h2o-public-test-data/smalldata/airlines/allyears2k_headers.zip")

# convert columns to factors
airlines["Year"]= airlines["Year"].asfactor()
airlines["Month"]= airlines["Month"].asfactor()
airlines["DayOfWeek"] = airlines["DayOfWeek"].asfactor()
airlines["Cancelled"] = airlines["Cancelled"].asfactor()
airlines['FlightNum'] = airlines['FlightNum'].asfactor()

# set the predictor names and the response column name
predictors = ["Origin", "Dest", "Year", "UniqueCarrier",
              "DayOfWeek", "Month", "Distance", "FlightNum"]
response = "IsDepDelayed"

# split into train val and test
train, valid, test = airlines.split_frame(ratios=[.8, .1], seed=1234)

Checking whether there is an H2O instance running at http://localhost:54321. connected.


--------------------------  -----------------------------
H2O_cluster_uptime:         8 mins 26 secs
H2O_cluster_timezone:       America/New_York
H2O_data_parsing_timezone:  UTC
H2O_cluster_version:        3.46.0.1
H2O_cluster_version_age:    24 days
H2O_cluster_name:           H2O_from_python_darien_b05g38
H2O_cluster_total_nodes:    1
H2O_cluster_free_memory:    7.925 Gb
H2O_cluster_total_cores:    12
H2O_cluster_allowed_cores:  12
H2O_cluster_status:         locked, healthy
H2O_connection_url:         http://localhost:54321
H2O_connection_proxy:       {"http": null, "https": null}
H2O_internal_security:      False
Python_version:             3.11.8 final
--------------------------  -----------------------------

Parse progress: |████████████████████████████████████████████████████████████████| (done) 100%


# 5.1.a

In [8]:
from h2o.grid.grid_search import H2OGridSearch

rf_estimator = estimators.H2ORandomForestEstimator()

ntrees = [10, 30, 50, 100]
max_depth = [1, 2, 4, 6]

hyper_parameters = {"ntrees": [10, 30, 50, 100], "max_depth": [1, 2, 4, 6]}

# Train the grid
rf_grid = H2OGridSearch(model=rf_estimator, hyper_params=hyper_parameters)
rf_grid.train(x=predictors, 
              y=response, 
              training_frame=train, 
              validation_frame=valid,
)


drf Grid Build progress: |

███████████████████████████████████████████████████████| (done) 100%


Hyper-Parameter Search Summary: ordered by increasing logloss
    max_depth    ntrees    model_ids                                                         logloss
--  -----------  --------  ----------------------------------------------------------------  ---------
    6            30        Grid_DRF_py_40_sid_a544_model_python_1712505565922_3059_model_8   0.611092
    6            100       Grid_DRF_py_40_sid_a544_model_python_1712505565922_3059_model_16  0.611462
    6            50        Grid_DRF_py_40_sid_a544_model_python_1712505565922_3059_model_12  0.611727
    6            10        Grid_DRF_py_40_sid_a544_model_python_1712505565922_3059_model_4   0.61557
    4            100       Grid_DRF_py_40_sid_a544_model_python_1712505565922_3059_model_15  0.626486
    4            50        Grid_DRF_py_40_sid_a544_model_python_1712505565922_3059_model_11  0.627005
    4            30        Grid_DRF_py_40_sid_a544_model_python_1712505565922_3059_model_7   0.627436
    4            10        Grid_DRF_py_40_sid_a544_model_python_1712505565922_3059_model_3   0.632004
    2            10        Grid_DRF_py_40_sid_a544_model_python_1712505565922_3059_model_2   0.650541
    2            50        Grid_DRF_py_40_sid_a544_model_python_1712505565922_3059_model_10  0.650695
    2            100       Grid_DRF_py_40_sid_a544_model_python_1712505565922_3059_model_14  0.651268
    2            30        Grid_DRF_py_40_sid_a544_model_python_1712505565922_3059_model_6   0.65209
    1            10        Grid_DRF_py_40_sid_a544_model_python_1712505565922_3059_model_1   0.665631
    1            50        Grid_DRF_py_40_sid_a544_model_python_1712505565922_3059_model_9   0.666897
    1            100       Grid_DRF_py_40_sid_a544_model_python_1712505565922_3059_model_13  0.66998
    1            30        Grid_DRF_py_40_sid_a544_model_python_1712505565922_3059_model_5   0.673536

# 5.1.b

In [91]:
# Get the grid results
grid_results = rf_grid.get_grid(sort_by='accuracy', decreasing=True)
print(grid_results)


Hyper-Parameter Search Summary: ordered by decreasing accuracy
    max_depth    ntrees    model_ids                                                         accuracy
--  -----------  --------  ----------------------------------------------------------------  ----------
    6            50        Grid_DRF_py_40_sid_a544_model_python_1712505565922_3059_model_12  0.671596
    6            30        Grid_DRF_py_40_sid_a544_model_python_1712505565922_3059_model_8   0.668075
    6            100       Grid_DRF_py_40_sid_a544_model_python_1712505565922_3059_model_16  0.66784
    6            10        Grid_DRF_py_40_sid_a544_model_python_1712505565922_3059_model_4   0.66784
    4            50        Grid_DRF_py_40_sid_a544_model_python_1712505565922_3059_model_11  0.657277
    4            100       Grid_DRF_py_40_sid_a544_model_python_1712505565922_3059_model_15  0.656338
    4            30        Grid_DRF_py_40_sid_a544_model_python_1712505565922_3059_model_7   0.655634
    4            10

# 5.1.c

In [94]:
best_model = grid_results.models[0]

performance = best_model.model_performance(test)

print(f"Best Model ID: {best_model.model_id}")
print(f"Accuracy: {performance.accuracy()[0][1]:.4f}")
print(f"AUC score: {performance.auc():.4f}")


Best Model ID: Grid_DRF_py_40_sid_a544_model_python_1712505565922_3059_model_12
Accuracy: 0.6781
AUC score: 0.7380


```text
Best Model ID: Grid_DRF_py_40_sid_a544_model_python_1712505565922_3059_model_12
Accuracy: 0.6781
AUC score: 0.7380
```

The best model in the grid search is a decision tree model with an accuracy of 0.6781 and an AUC score of 0.7380.

## 5.2 Randomized grid search

# 5.2.a

In [21]:
# %% [markdown]
# # 5.2.a

# %%


rf_estimator = estimators.H2ORandomForestEstimator()

ntrees = [10, 30, 50, 100]
max_depth = [1, 2, 4, 6]

hyper_parameters = {"ntrees": [10, 30, 50, 100], "max_depth": [1, 2, 4, 6]}

search_criteria = {'strategy': 'RandomDiscrete', 'max_models': 10, 'seed': 9232}

# Train the grid
rf_grid2 = H2OGridSearch(model=rf_estimator, hyper_params=hyper_parameters, search_criteria=search_criteria)
rf_grid2.train(x=predictors, 
              y=response, 
              training_frame=train, 
              validation_frame=valid,
)



rf_grid2.train(
    x=predictors,
    y=response,
    training_frame=train,
    validation_frame=valid,
)



drf Grid Build progress: |

███████████████████████████████████████████████████████| (done) 100%
drf Grid Build progress: |███████████████████████████████████████████████████████| (done) 100%
Hyper-Parameter Search Summary: ordered by decreasing accuracy
    max_depth    ntrees    model_ids                                                         accuracy
--  -----------  --------  ----------------------------------------------------------------  ----------
    6            100       Grid_DRF_py_40_sid_a544_model_python_1712505565922_6060_model_1   0.670423
    6            30        Grid_DRF_py_40_sid_a544_model_python_1712505565922_6060_model_9   0.665962
    4            50        Grid_DRF_py_40_sid_a544_model_python_1712505565922_6060_model_10  0.659624
    4            10        Grid_DRF_py_40_sid_a544_model_python_1712505565922_6060_model_4   0.655164
    4            30        Grid_DRF_py_40_sid_a544_model_python_1712505565922_6060_model_5   0.655164
    4            100       Grid_DRF_py_40_sid_a544_model_


# 5.2.b


In [22]:

# %%
# Get the random grid results
random_grid2_results = rf_grid2.get_grid(sort_by='accuracy', decreasing=True)
print(random_grid2_results)

Hyper-Parameter Search Summary: ordered by decreasing accuracy
    max_depth    ntrees    model_ids                                                         accuracy
--  -----------  --------  ----------------------------------------------------------------  ----------
    6            100       Grid_DRF_py_40_sid_a544_model_python_1712505565922_6060_model_1   0.670423
    6            30        Grid_DRF_py_40_sid_a544_model_python_1712505565922_6060_model_9   0.665962
    4            50        Grid_DRF_py_40_sid_a544_model_python_1712505565922_6060_model_10  0.659624
    4            10        Grid_DRF_py_40_sid_a544_model_python_1712505565922_6060_model_4   0.655164
    4            30        Grid_DRF_py_40_sid_a544_model_python_1712505565922_6060_model_5   0.655164
    4            100       Grid_DRF_py_40_sid_a544_model_python_1712505565922_6060_model_7   0.654695
    2            30        Grid_DRF_py_40_sid_a544_model_python_1712505565922_6060_model_3   0.64108
    2            1

# 3. H2O AutoML


#  5.3.a


In [23]:

# %%
from h2o.automl import H2OAutoML

# Initialize the AutoML object
aml = H2OAutoML(max_models=20, seed=9232)

# Train the AutoML
aml.train(x=predictors, y=response, training_frame=train, validation_frame=valid)


AutoML progress: |
12:39:44.888: User specified a validation frame with cross-validation still enabled. Please note that the models will still be validated using cross-validation only, the validation frame will be used to provide purely informative validation metrics on the trained models.
12:39:44.890: AutoML: XGBoost is not available; skipping it.

███████████████████████████████████████████████████████████████Job request failed Unexpected HTTP error: ('Connection aborted.', BadStatusLine('GET /3/Jobs/$03017f00000132d4ffffffff$_a460fe8ed666c9228a7e3806744456b HTTP/1.1\r\n')), will retry after 3s.
| (done) 100%


Model Details
=============
H2OStackedEnsembleEstimator : Stacked Ensemble
Model Key: StackedEnsemble_AllModels_1_AutoML_1_20240407_123944


Model Summary for Stacked Ensemble: 
key                                        value
-----------------------------------------  ----------------
Stacking strategy                          cross_validation
Number of base models (used / total)       3/20
# GBM base models (used / total)           0/10
# DRF base models (used / total)           1/2
# DeepLearning base models (used / total)  2/7
# GLM base models (used / total)           0/1
Metalearner algorithm                      GLM
Metalearner fold assignment scheme         Random
Metalearner nfolds                         5
Metalearner fold_column
Custom metalearner hyperparameters         None

ModelMetricsBinomialGLM: stackedensemble
** Reported on train data. **

MSE: 0.15537053190043382
RMSE: 0.39417068878904965
LogLoss: 0.48391873898936133
AUC: 0.9108567922000878
AUCPR: 0.9264786146223086
Gini: 0.8217135844001755
Null degrees of freedom: 9960
Residual degrees of freedom: 9957
Null deviance: 13768.661134226904
Residual deviance: 9640.629118146056
AIC: 9648.629118146056

Confusion Matrix (Act/Pred) for max f1 @ threshold = 0.49444672382040356
       NO    YES    Error    Rate
-----  ----  -----  -------  ---------------
NO     3485  1168   0.251    (1168.0/4653.0)
YES    667   4641   0.1257   (667.0/5308.0)
Total  4152  5809   0.1842   (1835.0/9961.0)

Maximum Metrics: Maximum metrics at their respective thresholds
metric                       threshold    value     idx
---------------------------  -----------  --------  -----
max f1                       0.494447     0.834937  226
max f2                       0.400938     0.897711  283
max f0point5                 0.574844     0.85184   181
max accuracy                 0.518754     0.819898  213
max precision                0.987661     1         0
max recall                   0.272308     1         341
max specificity              0.987661     1         0
max absolute_mcc             0.520367     0.638544  212
max min_per_class_accuracy   0.522363     0.818576  211
max mean_per_class_accuracy  0.520367     0.819591  212
max tns                      0.987661     4653      0
max fns                      0.987661     5295      0
max fps                      0.021916     4653      399
max tps                      0.272308     5308      341
max tnr                      0.987661     1         0
max fnr                      0.987661     0.997551  0
max fpr                      0.021916     1         399
max tpr                      0.272308     1         341

Gains/Lift Table: Avg response rate: 53.29 %, avg score: 54.17 %
group    cumulative_data_fraction    lower_threshold    lift        cumulative_lift    response_rate    score     cumulative_response_rate    cumulative_score    capture_rate    cumulative_capture_rate    gain      cumulative_gain    kolmogorov_smirnov
-------  --------------------------  -----------------  ----------  -----------------  ---------------  --------  --------------------------  ------------------  --------------  -------------------------  --------  -----------------  --------------------
1        0.0101395                   0.964276           1.8766      1.8766             1                0.980361  1                           0.980361            0.0190279       0.0190279                  87.6601   87.6601            0.0190279
2        0.0200783                   0.938184           1.8766      1.8766             1                0.949991  1                           0.965328            0.0186511       0.037679                   87.6601   87.6601            0.037679
3        0.0301175                   0.918946           1.8766      1.8766             1                0.929104  1                           0.953253            0.0188395       0.0565185                  87.6601   87.6601            0.0565185
4        0.0400562                   0.903226    

In [24]:
import pickle 
with open('aml.pkl', 'wb') as f:
    pickle.dump(aml, f)


# 5.3.b


In [60]:
# leaderboard
import pandas as pd 
pd.set_option('display.max_rows', None)
aml_lb = h2o.automl.get_leaderboard(aml)
aml_leaderboard = aml.leaderboard.sort(by='auc', ascending=False)
# display(aml_leaderboard.head(50))
print(aml_lb)

best_model = aml

model_id                                                      auc    logloss     aucpr    mean_per_class_error      rmse       mse
StackedEnsemble_AllModels_1_AutoML_1_20240407_123944     0.756087   0.585832  0.767857                0.348354  0.447694  0.20043
StackedEnsemble_BestOfFamily_1_AutoML_1_20240407_123944  0.755393   0.586386  0.767101                0.339776  0.447966  0.200674
GBM_1_AutoML_1_20240407_123944                           0.752938   0.58934   0.76335                 0.351937  0.449191  0.201773
GBM_3_AutoML_1_20240407_123944                           0.751078   0.590976  0.760174                0.347079  0.449897  0.202407
GBM_2_AutoML_1_20240407_123944                           0.749332   0.592415  0.759057                0.353812  0.450614  0.203053
GBM_4_AutoML_1_20240407_123944                           0.746035   0.596445  0.755082                0.357029  0.452407  0.204672
GBM_5_AutoML_1_20240407_123944                           0.745896   0.595113  0.7556

### Best AML Model and AUC

In [80]:
from pprint import pprint 
best_model = aml.leader

print(f"Best model: {best_model.model_id}")
print(f"AUC score: {best_model.auc():.4f}")



Best model: StackedEnsemble_AllModels_1_AutoML_1_20240407_123944
AUC score: 0.9109


### Full Parameters

In [84]:
display(best_model)
display(best_model.params)

Model Details
=============
H2OStackedEnsembleEstimator : Stacked Ensemble
Model Key: StackedEnsemble_AllModels_1_AutoML_1_20240407_123944


Model Summary for Stacked Ensemble: 
key                                        value
-----------------------------------------  ----------------
Stacking strategy                          cross_validation
Number of base models (used / total)       3/20
# GBM base models (used / total)           0/10
# DRF base models (used / total)           1/2
# DeepLearning base models (used / total)  2/7
# GLM base models (used / total)           0/1
Metalearner algorithm                      GLM
Metalearner fold assignment scheme         Random
Metalearner nfolds                         5
Metalearner fold_column
Custom metalearner hyperparameters         None

ModelMetricsBinomialGLM: stackedensemble
** Reported on train data. **

MSE: 0.15537053190043382
RMSE: 0.39417068878904965
LogLoss: 0.48391873898936133
AUC: 0.9108567922000878
AUCPR: 0.9264786146223086
Gini: 0.8217135844001755
Null degrees of freedom: 9960
Residual degrees of freedom: 9957
Null deviance: 13768.661134226904
Residual deviance: 9640.629118146056
AIC: 9648.629118146056

Confusion Matrix (Act/Pred) for max f1 @ threshold = 0.49444672382040356
       NO    YES    Error    Rate
-----  ----  -----  -------  ---------------
NO     3485  1168   0.251    (1168.0/4653.0)
YES    667   4641   0.1257   (667.0/5308.0)
Total  4152  5809   0.1842   (1835.0/9961.0)

Maximum Metrics: Maximum metrics at their respective thresholds
metric                       threshold    value     idx
---------------------------  -----------  --------  -----
max f1                       0.494447     0.834937  226
max f2                       0.400938     0.897711  283
max f0point5                 0.574844     0.85184   181
max accuracy                 0.518754     0.819898  213
max precision                0.987661     1         0
max recall                   0.272308     1         341
max specificity              0.987661     1         0
max absolute_mcc             0.520367     0.638544  212
max min_per_class_accuracy   0.522363     0.818576  211
max mean_per_class_accuracy  0.520367     0.819591  212
max tns                      0.987661     4653      0
max fns                      0.987661     5295      0
max fps                      0.021916     4653      399
max tps                      0.272308     5308      341
max tnr                      0.987661     1         0
max fnr                      0.987661     0.997551  0
max fpr                      0.021916     1         399
max tpr                      0.272308     1         341

Gains/Lift Table: Avg response rate: 53.29 %, avg score: 54.17 %
group    cumulative_data_fraction    lower_threshold    lift        cumulative_lift    response_rate    score     cumulative_response_rate    cumulative_score    capture_rate    cumulative_capture_rate    gain      cumulative_gain    kolmogorov_smirnov
-------  --------------------------  -----------------  ----------  -----------------  ---------------  --------  --------------------------  ------------------  --------------  -------------------------  --------  -----------------  --------------------
1        0.0101395                   0.964276           1.8766      1.8766             1                0.980361  1                           0.980361            0.0190279       0.0190279                  87.6601   87.6601            0.0190279
2        0.0200783                   0.938184           1.8766      1.8766             1                0.949991  1                           0.965328            0.0186511       0.037679                   87.6601   87.6601            0.037679
3        0.0301175                   0.918946           1.8766      1.8766             1                0.929104  1                           0.953253            0.0188395       0.0565185                  87.6601   87.6601            0.0565185
4        0.0400562                   0.903226    

{'model_id': {'default': None,
  'actual': {'__meta': {'schema_version': 3,
    'schema_name': 'ModelKeyV3',
    'schema_type': 'Key<Model>'},
   'name': 'StackedEnsemble_AllModels_1_AutoML_1_20240407_123944',
   'type': 'Key<Model>',
   'URL': '/3/Models/StackedEnsemble_AllModels_1_AutoML_1_20240407_123944'},
  'input': None},
 'training_frame': {'default': None,
  'actual': {'__meta': {'schema_version': 3,
    'schema_name': 'FrameKeyV3',
    'schema_type': 'Key<Frame>'},
   'name': 'AutoML_1_20240407_123944_training_py_40_sid_a544',
   'type': 'Key<Frame>',
   'URL': '/3/Frames/AutoML_1_20240407_123944_training_py_40_sid_a544'},
  'input': {'__meta': {'schema_version': 3,
    'schema_name': 'FrameKeyV3',
    'schema_type': 'Key<Frame>'},
   'name': 'AutoML_1_20240407_123944_training_py_40_sid_a544',
   'type': 'Key<Frame>',
   'URL': '/3/Frames/AutoML_1_20240407_123944_training_py_40_sid_a544'}},
 'response_column': {'default': None,
  'actual': {'__meta': {'schema_version': 3,
    


# 5.3.c


In [85]:
# * (c) Display the AUC score of the best model for the test set. (2)

performance = best_model.model_performance(test)
print(f"AUC score (test set): {performance.auc():.4f}")


AUC score (test set): 0.7452


# 5.3.d

No extreme gradient boosting models were trained by the AutoML model. Therefore, I have adjusted part d to identify the best gradient boosting model.

In [90]:
# * (d) Identify the best XGBoost model among all the models tested using log loss as the criteria. (2)

aml_lb = h2o.automl.get_leaderboard(aml, extra_columns = 'ALL').sort(by='logloss', ascending=True)
display(aml_lb.head(49))

# xgb = aml.get_best_model(algorithm="xgboost", criterion="logloss")
gbm = aml.get_best_model(algorithm="gbm", criterion="logloss")

print(f"Best gbm model: {gbm.model_id}")
print(f"Log loss: {gbm.logloss():.4f}")

model_id                                                      auc    logloss     aucpr    mean_per_class_error      rmse       mse    training_time_ms    predict_time_per_row_ms  algo
-------------------------------------------------------  --------  ---------  --------  ----------------------  --------  --------  ------------------  -------------------------  ---------------
StackedEnsemble_AllModels_1_AutoML_1_20240407_123944     0.756087   0.585832  0.767857                0.348354  0.447694  0.20043                10583                   0.119953  StackedEnsemble
StackedEnsemble_BestOfFamily_1_AutoML_1_20240407_123944  0.755393   0.586386  0.767101                0.339776  0.447966  0.200674                6984                   0.110318  StackedEnsemble
GBM_1_AutoML_1_20240407_123944                           0.752938   0.58934   0.76335                 0.351937  0.449191  0.201773                 876                   0.011968  GBM
GBM_3_AutoML_1_20240407_123944                           0.751078   0.590976  0.760174                0.347079  0.449897  0.202407                 527                   0.010315  GBM
GBM_2_AutoML_1_20240407_123944                           0.749332   0.592415  0.759057                0.353812  0.450614  0.203053                 490                   0.01101   GBM
GBM_5_AutoML_1_20240407_123944                           0.745896   0.595113  0.755672                0.355188  0.451947  0.204256                 434                   0.010734  GBM
GBM_4_AutoML_1_20240407_123944                           0.746035   0.596445  0.755082                0.357029  0.452407  0.204672                 857                   0.010363  GBM
GBM_grid_1_AutoML_1_20240407_123944_model_2              0.743288   0.596842  0.754765                0.362023  0.45291   0.205128                 385                   0.011329  GBM
GBM_grid_1_AutoML_1_20240407_123944_model_1              0.741982   0.598302  0.75391                 0.359084  0.453623  0.205774                 535                   0.014209  GBM
XRT_1_AutoML_1_20240407_123944                           0.743518   0.602369  0.750886                0.35747   0.455421  0.207408                3657                   0.020476  DRF
GLM_1_AutoML_1_20240407_123944                           0.735386   0.604103  0.748049                0.361836  0.456438  0.208336                2015                   0.000625  GLM
GBM_grid_1_AutoML_1_20240407_123944_model_3              0.73383    0.604808  0.745791                0.381682  0.45677   0.208639                 369                   0.009421  GBM
DeepLearning_grid_1_AutoML_1_20240407_123944_model_1     0.738192   0.607502  0.752143                0.367511  0.456748  0.208618              375854                   0.111428  DeepLearning
GBM_grid_1_AutoML_1_20240407_123944_model_4              0.739077   0.608325  0.749779                0.356197  0.457472  0.209281                1303                   0.011094  GBM
GBM_grid_1_AutoML_1_20240407_123944_model_5              0.737314   0.610323  0.747433                0.365736  0.458112  0.209867                1403                   0.010514  GBM
DeepLearning_1_AutoML_1_20240407_123944                  0.726267   0.611095  0.740343                0.383504  0.459766  0.211385                2068                   0.014683  DeepLearning
DeepLearning_grid_3_AutoML_1_20240407_123944_model_2     0.725047   0.613322  0.738721                0.375088  0.460784  0.212322               67908                   0.02575   DeepLearning
DeepLearning_grid_2_AutoML_1_20240407_123944_model_1     0.730893   0.614709  0.742126                0.368014  0.45942   0.211066              385478                   0.116829  DeepLearning
DRF_1_AutoML_1_20240407_123944                           0.736245   0.616897  0.746916                0.365627  0.459477  0.211119                3265                   0.023223  DRF
DeepLearning_grid_2_AutoML_1_20240407_123944_model_2     0.725957   0.619504  0.7372

Best gbm model: GBM_1_AutoML_1_20240407_123944
Log loss: 0.5104


As mentioned above, no extreme gradient boosting models were trained by the AutoML model, so gradient boosting models are considered in-place. The best gradient boosting model with respect to log loss is: `GBM_1_AutoML_1_20240407_123944`
```text
Best gbm model: GBM_1_AutoML_1_20240407_123944
Log loss: 0.5104
```
